src/core/reactor_backend.cc

```diff
@@ -1475,46 +1511,39 @@ class reactor_backend_uring final : public reactor_backend {
     }
     virtual future<size_t> read_some(pollable_fd_state& fd, void* buffer, size_t len) override {
         return readable(fd).then([this, &fd, buffer, len] () mutable {
-            auto r = fd.fd.read(buffer, len);
-            if (!r) {
-                return read_some(fd, buffer, len);
-            }
-            if (size_t(*r) == len) {
+            return queue_request(internal::io_request::make_read(fd.fd.get(), -1, buffer, len, false));
+        }).then([&fd, len] (size_t result) {
+            if (result == len) {
                 fd.speculate_epoll(EPOLLIN);
             }
-            return make_ready_future<size_t>(*r);
+            return make_ready_future<size_t>(result);
         });
     }
     virtual future<size_t> read_some(pollable_fd_state& fd, const std::vector<iovec>& iov) override {
-        return readable(fd).then([this, &fd, iov = iov] () mutable {
-            ::msghdr mh = {};
-            mh.msg_iov = &iov[0];
-            mh.msg_iovlen = iov.size();
-            auto r = fd.fd.recvmsg(&mh, 0);
-            if (!r) {
-                return read_some(fd, iov);
-            }
-            if (size_t(*r) == internal::iovec_len(iov)) {
-                fd.speculate_epoll(EPOLLIN);
-            }
-            return make_ready_future<size_t>(*r);
+        return do_with(::msghdr{}, iov, [&fd, this] (auto& mh, auto& iov) {
```

Comment:
If you have a specialized io_completion subclass, it can host the msghdr so you can avoid the extra allocation.